python/python/lance/fragment.py

```diff
@@ -37,53 +36,161 @@
-    def to_json(self) -> str:
-        """Serialize :class:`FragmentMetadata` to a JSON blob"""
-        return json.loads(self._metadata.json())
+    def data_files(self) -> List[DataFile]:
+        warnings.warn(
+            "FragmentMetadata.data_files is deprecated. Use .files instead.",
+            DeprecationWarning,
+        )
+        return self.files
+
+    def to_json(self) -> dict:
+        """Get this as a simple JSON-serializable dictionary."""
+        files = [asdict(f) for f in self.files]
+        for f in files:
+            f["path"] = f.pop("_path")
+        return dict(
+            id=self.id,
+            files=files,
+            physical_rows=self.physical_rows,
+            deletion_file=(
+                self.deletion_file.asdict() if self.deletion_file is not None else None
+            ),
+            row_id_meta=(
+                self.row_id_meta.asdict() if self.row_id_meta is not None else None
+            ),
+        )
 
     @staticmethod
     def from_json(json_data: str) -> FragmentMetadata:
-        """Reconstruct :class:`FragmentMetadata` from a JSON blob"""
-        return FragmentMetadata(json_data)
+        json_data = json.loads(json_data)
+
+        deletion_file = json_data.get("deletion_file")
+        if deletion_file is not None:
+            deletion_file = DeletionFile(**deletion_file)
+
+        row_id_meta = json_data.get("row_id_meta")
+        if row_id_meta is not None:
+            row_id_meta = RowIdMeta(**row_id_meta)
+
+        return FragmentMetadata(
+            id=json_data["id"],
+            files=[DataFile(**f) for f in json_data["files"]],
+            physical_rows=json_data["physical_rows"],
+            deletion_file=deletion_file,
+            row_id_meta=row_id_meta,
+        )
 
-    def data_files(self) -> Iterable[str]:
-        """Return the data files of the fragment"""
-        return self._metadata.data_files()
 
-    def deletion_file(self):
-        """Return the deletion file, if any"""
-        return self._metadata.deletion_file()
+@dataclass
+class DataFile:
+    """
+    A data file in a fragment.
+
+    Attributes
+    ----------
+    path : str
+        The path to the data file.
+    fields : List[int]
+        The field ids of the columns in this file.
+    column_indices : List[int]
+        The column indices in the original schema.
+    file_major_version : int
+        The major version of the data storage format.
+    file_minor_version : int
+        The minor version of the data storage format.
+    """
+
+    _path: str
+    fields: List[int]
+    column_indices: List[int] = field(default_factory=list)
+    file_major_version: int = 0
+    file_minor_version: int = 0
+
+    def __init__(
+        self,
+        path: str,
+        fields: List[int],
+        column_indices: List[int] = None,
+        file_major_version: int = 0,
+        file_minor_version: int = 0,
+    ):
+        # TODO: only we eliminate the path method, we can remove this
```

Comment:
```suggestion
        # TODO: once we eliminate the path method, we can remove this
```
Maybe?

What does it mean to eliminate the path method?

Reply:
Previously `path` was a method `DataFile.path()`. But now, it's a property: `DataFile.path`. I go through a lot of hoops to support both, but eventually I'd like to remove the method support, so it's only accessible as a property. Then all of this code will be much more straightforward.